core/src/main/java/org/testcontainers/utility/ResourceReaper.java

```diff
@@ -1,37 +1,164 @@
 import com.github.dockerjava.api.exception.InternalServerErrorException;
 import com.github.dockerjava.api.exception.NotFoundException;
+import com.github.dockerjava.api.model.AccessMode;
+import com.github.dockerjava.api.model.Bind;
+import com.github.dockerjava.api.model.ExposedPort;
+import com.github.dockerjava.api.model.HostConfig;
 import com.github.dockerjava.api.model.Network;
+import com.github.dockerjava.api.model.Ports;
+import com.github.dockerjava.api.model.Volume;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.http.client.utils.URLEncodedUtils;
+import org.apache.http.message.BasicNameValuePair;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.testcontainers.DockerClientFactory;
 
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.Socket;
+import java.util.AbstractMap.SimpleEntry;
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 /**
  * Component that responsible for container removal and automatic cleanup of dead containers at JVM shutdown.
  */
+@Slf4j
 public final class ResourceReaper {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(ResourceReaper.class);
+
+    private static final List<List<Map.Entry<String, String>>> DEATH_NOTE = new ArrayList<>();
+
     private static ResourceReaper instance;
     private final DockerClient dockerClient;
     private Map<String, String> registeredContainers = new ConcurrentHashMap<>();
     private Set<String> registeredNetworks = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    private AtomicBoolean hookIsSet = new AtomicBoolean(false);
 
     private ResourceReaper() {
         dockerClient = DockerClientFactory.instance().client();
+    }
+
+    @SneakyThrows(InterruptedException.class)
+    public static String start(String hostIpAddress, DockerClient client) {
+        String ryukImage = TestcontainersConfiguration.getInstance().getRyukImage();
+        DockerClientFactory.instance().checkAndPullImage(client, ryukImage);
+
+        MountableFile mountableFile = MountableFile.forClasspathResource(ResourceReaper.class.getName().replace(".", "/") + ".class");
+
+        String ryukContainerId = client.createContainerCmd(ryukImage)
+                .withHostConfig(new HostConfig() {
+                    @JsonProperty("AutoRemove")
+                    boolean autoRemove = true;
+                })
+                .withExposedPorts(new ExposedPort(8080))
+                .withPublishAllPorts(true)
+                .withName("tc-ryuk-" + DockerClientFactory.SESSION_ID)
```

Comment:
Nice to name the container 😄 
Please could we just expand `tc` to `testcontainers` though? It's just a thing that people can google more easily if they're wondering what this container is.